src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -21,127 +24,137 @@
-      String wishlist = Jsoup.clean(request.getParameter("wishlist"), Whitelist.none());
-      String teachlist = Jsoup.clean(request.getParameter("teachlist"), Whitelist.none());
-      String id = Jsoup.clean(request.getParameter("id"), Whitelist.none());
-
-      // add kind users to keyfactory
-      keyFactory.setKind("users");
-
-      // take the user elements and create a new entity object for it 
-      FullEntity userEntity =
-        Entity.newBuilder(keyFactory.newKey(id))
-            .set("name", name)
-            .set("email", email)
-            .set("wishlist", wishlist)
-            .set("teachlist", teachlist)
-            .build();
-      
-      // insert that new entity
-      try{
-        datastore.add(userEntity);
-      }
-      catch(Exception DatastoreException){
-        this.doPut(request, response);
-        return;
-      }
-      response.sendRedirect("index.html");
-    }
+  // create new datastore and keyfactory
+  Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+  KeyFactory keyFactory = datastore.newKeyFactory();
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException { // create
 
-    @Override
-    public void doPut(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException  { // update
-        
-      // get form information
-      String NOTHING = Jsoup.clean(request.getParameter("NOTHING"), Whitelist.none());
-      String name = Jsoup.clean(request.getParameter("name"), Whitelist.none());
-      String email = Jsoup.clean(request.getParameter("email"), Whitelist.none());
-      String wishlist = Jsoup.clean(request.getParameter("wishlist"), Whitelist.none());
-      String teachlist = Jsoup.clean(request.getParameter("teachlist"), Whitelist.none());
-      String id = Jsoup.clean(request.getParameter("id"), Whitelist.none());
-
-      //access the original entity
-      keyFactory.setKind("users");
-      Key userKey = keyFactory.newKey(id);
-      Entity user = datastore.get(userKey);
-
-      if(wishlist.equals(NOTHING)){
-          wishlist = user.getString("wishlist");
-      }
-      if(teachlist.equals(NOTHING)){
-          teachlist = user.getString("teachlist");
-      }
-
-      // take the user elements and create a new updated entity object for it 
-      FullEntity userUpdatedEntity =
-        Entity.newBuilder(user)
-            .set("name", name)
-            .set("email", email)
-            .set("wishlist", wishlist)
-            .set("teachlist", teachlist)
-            .build();
-
-      //update the information in the original entity
-      datastore.put(userUpdatedEntity);
-      response.sendRedirect("index.html");
+    String button_type = request.getParameter("submission").toString();
+    if (button_type == "update") {
+      this.doPut(request, response);
       return;
     }
```

Comment:
Gah, I think I may have confused you cuz I hadn't thought of the whole picture. POST is used for creates **only**. Not replaces and not updates. Also PUT is used for create/replace **only.** If the client is replacing an existing user, then it should be calling PUT /users **not** POST. POST should fail for existing users.

```suggestion
```

Reply:
ah ok, gotcha, i'll separate them now